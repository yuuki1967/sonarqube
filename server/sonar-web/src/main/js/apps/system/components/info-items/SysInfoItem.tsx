/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { map } from 'lodash';
import HealthItem from './HealthItem';
import AlertErrorIcon from '../../../../components/icons-components/AlertErrorIcon';
import AlertSuccessIcon from '../../../../components/icons-components/AlertSuccessIcon';
import { HealthType, SysValue, SysValueObject } from '../../../../api/system';
import { HEALTH_FIELD } from '../../utils';

interface Props {
  name: string;
  value: SysValue;
}

export default function SysInfoItem({ name, value }: Props): JSX.Element {
  if (name === HEALTH_FIELD || name === 'State') {
    return <HealthItem className="no-margin" health={value as HealthType} />;
  }
  if (value instanceof Array) {
    return <code>{JSON.stringify(value)}</code>;
  }
  switch (typeof value) {
    case 'boolean':
      return <BooleanItem value={value as boolean} />;
    case 'object':
      return <ObjectItem value={value as SysValueObject} />;
    default:
      return <code>{value}</code>;
  }
}

function BooleanItem({ value }: { value: boolean }) {
  if (value) {
    return <AlertSuccessIcon />;
  } else {
    return <AlertErrorIcon />;
  }
}

function ObjectItem({ value }: { value: SysValueObject }) {
  return (
    <table className="data">
      <tbody>
        {map(value, (value, name) => (
          <tr key={name}>
            <td className="thin nowrap">{name}</td>
            <td>
              <SysInfoItem name={name} value={value} />
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  );
}
