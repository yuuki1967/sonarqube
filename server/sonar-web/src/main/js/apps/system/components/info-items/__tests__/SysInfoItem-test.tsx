/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow, mount } from 'enzyme';
import SysInfoItem from '../SysInfoItem';

it('should render string', () => {
  const wrapper = shallow(<SysInfoItem name="foo" value="/some/path/as/an/example" />);
  expect(wrapper.find('code').text()).toBe('/some/path/as/an/example');
});

it('should render object', () => {
  const wrapper = shallow(<SysInfoItem name="foo" value={{ bar: 'baz' }} />);
  expect(wrapper.find('ObjectItem').prop('value')).toEqual({ bar: 'baz' });
});

it('should render boolean', () => {
  const wrapper = shallow(<SysInfoItem name="foo" value={true} />);
  expect(wrapper.find('BooleanItem').prop('value')).toBe(true);
});

it('should render health item', () => {
  const wrapper = shallow(<SysInfoItem name="Health" value="GREEN" />);
  expect(wrapper.find('HealthItem').prop('health')).toBe('GREEN');
});

it('should render object correctly', () => {
  expect(
    mount(
      <SysInfoItem name="test" value={{ foo: 'Far', bar: { a: 1, b: 'b' }, baz: true }} />
    ).find('ObjectItem')
  ).toMatchSnapshot();
});

it('should render `true`', () => {
  const wrapper = mount(<SysInfoItem name="test" value={true} />);
  expect(wrapper.find('AlertSuccessIcon').exists()).toBeTruthy();
});

it('should render `false`', () => {
  const wrapper = mount(<SysInfoItem name="test" value={false} />);
  expect(wrapper.find('AlertErrorIcon').exists()).toBeTruthy();
});
