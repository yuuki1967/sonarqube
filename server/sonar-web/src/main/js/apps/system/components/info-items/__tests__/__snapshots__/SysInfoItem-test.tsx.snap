// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render object correctly 1`] = `
Array [
  <ObjectItem
    value={
      Object {
        "bar": Object {
          "a": 1,
          "b": "b",
        },
        "baz": true,
        "foo": "Far",
      }
    }
  >
    <table
      className="data"
    >
      <tbody>
        <tr
          key="foo"
        >
          <td
            className="thin nowrap"
          >
            foo
          </td>
          <td>
            <SysInfoItem
              name="foo"
              value="Far"
            >
              <code>
                Far
              </code>
            </SysInfoItem>
          </td>
        </tr>
        <tr
          key="bar"
        >
          <td
            className="thin nowrap"
          >
            bar
          </td>
          <td>
            <SysInfoItem
              name="bar"
              value={
                Object {
                  "a": 1,
                  "b": "b",
                }
              }
            >
              <ObjectItem
                value={
                  Object {
                    "a": 1,
                    "b": "b",
                  }
                }
              >
                <table
                  className="data"
                >
                  <tbody>
                    <tr
                      key="a"
                    >
                      <td
                        className="thin nowrap"
                      >
                        a
                      </td>
                      <td>
                        <SysInfoItem
                          name="a"
                          value={1}
                        >
                          <code>
                            1
                          </code>
                        </SysInfoItem>
                      </td>
                    </tr>
                    <tr
                      key="b"
                    >
                      <td
                        className="thin nowrap"
                      >
                        b
                      </td>
                      <td>
                        <SysInfoItem
                          name="b"
                          value="b"
                        >
                          <code>
                            b
                          </code>
                        </SysInfoItem>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </ObjectItem>
            </SysInfoItem>
          </td>
        </tr>
        <tr
          key="baz"
        >
          <td
            className="thin nowrap"
          >
            baz
          </td>
          <td>
            <SysInfoItem
              name="baz"
              value={true}
            >
              <BooleanItem
                value={true}
              >
                <AlertSuccessIcon>
                  <Icon>
                    <svg
                      height={16}
                      style={
                        Object {
                          "clipRule": "evenodd",
                          "fillRule": "evenodd",
                          "strokeLinejoin": "round",
                          "strokeMiterlimit": 1.41421,
                        }
                      }
                      version="1.1"
                      viewBox="0 0 16 16"
                      width={16}
                      xmlSpace="preserve"
                      xmlnsXlink="http://www.w3.org/1999/xlink"
                    >
                      <path
                        d="M12.607 6.554q0-0.25-0.161-0.411l-0.813-0.804q-0.17-0.17-0.402-0.17t-0.402 0.17l-3.643 3.634-2.018-2.018q-0.17-0.17-0.402-0.17t-0.402 0.17l-0.813 0.804q-0.161 0.161-0.161 0.411 0 0.241 0.161 0.402l3.232 3.232q0.17 0.17 0.402 0.17 0.241 0 0.411-0.17l4.848-4.848q0.161-0.161 0.161-0.402zM14.857 8q0 1.866-0.92 3.442t-2.496 2.496-3.442 0.92-3.442-0.92-2.496-2.496-0.92-3.442 0.92-3.442 2.496-2.496 3.442-0.92 3.442 0.92 2.496 2.496 0.92 3.442z"
                        style={
                          Object {
                            "fill": "#00aa00",
                          }
                        }
                      />
                    </svg>
                  </Icon>
                </AlertSuccessIcon>
              </BooleanItem>
            </SysInfoItem>
          </td>
        </tr>
      </tbody>
    </table>
  </ObjectItem>,
  <ObjectItem
    value={
      Object {
        "a": 1,
        "b": "b",
      }
    }
  >
    <table
      className="data"
    >
      <tbody>
        <tr
          key="a"
        >
          <td
            className="thin nowrap"
          >
            a
          </td>
          <td>
            <SysInfoItem
              name="a"
              value={1}
            >
              <code>
                1
              </code>
            </SysInfoItem>
          </td>
        </tr>
        <tr
          key="b"
        >
          <td
            className="thin nowrap"
          >
            b
          </td>
          <td>
            <SysInfoItem
              name="b"
              value="b"
            >
              <code>
                b
              </code>
            </SysInfoItem>
          </td>
        </tr>
      </tbody>
    </table>
  </ObjectItem>,
]
`;
